pkg/logql/syntax/ast.go

```diff
@@ -92,6 +92,56 @@ func (m MultiStageExpr) stages() ([]log.Stage, error) {
 	return c, nil
 }
 
+// reorderStages reorders m such that LineFilters
+// are as close to the front of the filter as possible.
+func (m MultiStageExpr) reorderStages() []StageExpr {
+	var (
+		result  = make([]StageExpr, 0, len(m))
+		filters = make([]*LineFilterExpr, 0, len(m))
+		rest    = make([]StageExpr, 0, len(m))
+	)
+
+	for _, s := range m {
+		switch f := s.(type) {
+		case *LineFilterExpr:
+			filters = append(filters, f)
+		case *LineFmtExpr:
+			rest = append(rest, f)
+
+			if len(filters) > 0 {
+				result = append(result, combineFilters(filters))
+			}
+			result = append(result, rest...)
+
+			filters = filters[:0]
+			rest = rest[:0]
+		default:
+			rest = append(rest, f)
+		}
+	}
```

Comment:
I don't quite understand what we want to happen here. I do sort of because of the test, but it's not obvious to me from this why we'd want to put LineFilters at the front of each section of a query and not just all at the front in general. 

In the case of your test, it confirms my first guess that after certain parts of a logql expression we need to start the building of `filters` and `rest` again, but the reasoning is unclear.

Can we include a comment as to why the ordering needs to preserved to a certain extent? For example, why is the correct result
```
|= "foo" |= "next" |= "bar" |= "baz" | logfmt | line_format "{{.foo}}" |= "1" |= "2" |= "3" | logfmt
```

and not
```
|= "foo" |= "next" |= "bar" |= "baz" | logfmt |= "1" |= "2" |= "3" | line_format "{{.foo}}" |logfmt
```

Reply:
a `LineFormat` stage changes the line so any line filters that happen after it have to look at the new line instead.

example:
with the log line `foo=bar msg=something` and the logql expression `{} | line_format "new line" |= "new"`, we expect `|= "new"` to match but it won't if we put it before a line filter

Happy to add a comment

Reply:
Thanks, that makes sense. My lack of `logql` knowledge is showing.

I think a comment
```
// A line format expr will modify the contents of the line, so any line filter expressions that were originally after the line format expr must still be after the line format expr.
```
will help future readers understand.

Reply:
done 🙂